src/librustc_typeck/check/mod.rs

```diff
@@ -1709,23 +1710,184 @@ fn check_opaque_for_cycles<'tcx>(
     span: Span,
     origin: &hir::OpaqueTyOrigin,
 ) {
-    if let Err(partially_expanded_type) = tcx.try_expand_impl_trait_type(def_id.to_def_id(), substs)
-    {
-        if let hir::OpaqueTyOrigin::AsyncFn = origin {
-            struct_span_err!(tcx.sess, span, E0733, "recursion in an `async fn` requires boxing",)
+    if let Err(partially_expanded_type) = tcx.try_expand_impl_trait_type(def_id.to_def_id(), substs) {
+        match (origin, &partially_expanded_type.kind) {
+            (hir::OpaqueTyOrigin::AsyncFn, _) => {
+                struct_span_err!(
+                    tcx.sess,
+                    span,
+                    E0733,
+                    "recursion in an `async fn` requires boxing"
+                )
                 .span_label(span, "recursive `async fn`")
                 .note("a recursive `async fn` must be rewritten to return a boxed `dyn Future`")
                 .emit();
-        } else {
-            let mut err =
-                struct_span_err!(tcx.sess, span, E0720, "opaque type expands to a recursive type",);
-            err.span_label(span, "expands to a recursive type");
-            if let ty::Opaque(..) = partially_expanded_type.kind {
-                err.note("type resolves to itself");
-            } else {
-                err.note(&format!("expanded type is `{}`", partially_expanded_type));
             }
-            err.emit();
+            (hir::OpaqueTyOrigin::FnReturn, ty::Opaque(..)) => {
+                let mut err = struct_span_err!(
+                    tcx.sess,
+                    span,
+                    E0720,
+                    "cannot resolve opaque return type `{}` to a concrete type",
+                    partially_expanded_type
+                );
+                err.span_label(span, "cannot resolve to a concrete type");
+                if let Some((hir_id, body_id)) = tcx
```

Comment:
some comments would be nice :)

Reply:
e.g., "find the `hir_id` and `body_id` of the function that declared this `-> impl Trait` type" or something